WL-4250 Users can change the password in Accounts tool.

ouit0408 opened the pull request:
Now no two external users can have same email address, kept the unique
email functionality as 'password-rest' tool will not work otherwise.
In  UserAction class checking for email change by user before actually
checking for existing account with the new email.
Similarly, creating validation account only if email is changed to a new
unique email address.

buckett commented on user-tool/tool/src/java/org/sakaiproject/user/tool/UsersAction.java:
```diff
@@ -806,12 +806,15 @@ public void doSave(RunData data, Context context)
+				if (StringUtils.trimToNull(valueEmail) != null && StringUtils.trimToNull(oldEmail) != null && !(oldEmail.equals(valueEmail))
+						&& EmailValidator.getInstance().isValid(edit.getEid()) && !edit.getEid().equalsIgnoreCase(valueEmail)) {
 					validationLogic.createValidationAccount(edit.getId(),valueEmail);
 					addAlert(state,rb.getFormattedMessage("useedi.val.email",new String[]{valueEmail}));
```
Don't we want a return here so we don't save the user object?

buckett replied:
Correct, no I don't think we do.

ouit0408 replied:
we still need to save the user object because there could be other changes like firstname,password,etc

buckett commented on user-tool/tool/src/java/org/sakaiproject/user/tool/UsersAction.java:
```diff
@@ -1118,17 +1121,18 @@ else if ((mode != null) && (mode.equalsIgnoreCase("edit")) && (!singleUser))
-					addAlert(state,rb.getString("useedi.email.exists"));
-					return false;
-				}
+				addAlert(state,rb.getString("useedi.email.exists"));
```
This also prevents admins from changing a user's email address to one that is already used doesn't it?

ouit0408 replied:
yes it does as in both contexts same method is called.

ouit0408 replied:
Have added adminUser check so now no check will be done when admin changes email address of any user account. No alert is displayed for admin. Whereas other users cannot change their email to an existing one. have pushed the changes.

ouit0408 committed:
WL-4250 Users can change the password in Accounts tool.

Now no two external users can have same email address, kept the unique
email functionality as 'password-rest' tool will not work otherwise.
In  UserAction class checking for email change by user before actually
checking for existing account with the new email.
Similarly, creating validation account only if email is changed to a new
unique email address.
No 'validation creation' and 'email exists' check will be performed if
user is admin.

buckett commented on user-tool/tool/src/java/org/sakaiproject/user/tool/UsersAction.java:
```diff
@@ -806,12 +806,15 @@ public void doSave(RunData data, Context context)
-				String valueEmail = (String)state.getAttribute("valueEmail");
-				if (StringUtils.trimToNull(valueEmail) != null && EmailValidator.getInstance().isValid(edit.getEid()) && !edit.getEid().equalsIgnoreCase(valueEmail)) {
+				//start this validation only when user has changed the email for the account else skip, also skip for admin user
+				if (!SecurityService.isSuperUser() && StringUtils.trimToNull(valueEmail) != null && StringUtils.trimToNull(oldEmail) != null && !(oldEmail.equals(valueEmail))
```
I don't think you need this check as oldEmail will never be `null` because of your check in `readUserForm()`. Don't think it needs changing.

The complexity of this validation code should be refactored to make it all much clearer, but again not something for now.

ouit0408 replied:
oldEmail can be null as it is set in the state only when user has changed his email otherwise it will be null.

ouit0408 replied:
though i think valueEmail will never be null.
